### PR TITLE
Byron mainnet converter and validator

### PR DESCRIPTION
**Pull request by nc6**
This PR adds a standalone mainnet converter and validator tool, along with adding some additional tracing to the immutable DB.

Example usage:

```
[nc@lorien:~/topos/iohk/ouroboros-network]$ ./ouroboros-consensus/.stack-work/dist/x86_64-linux-nix/Cabal-2.4.0.1/build/byron-db-converter/byron-db-converter c
onvert --epochDir /tmp/tmp.rquJ7pBVhW --dbDir $(mktemp -d) --epochSlots 21600
Writing to "/tmp/tmp.hIWEfz5t3r/"
Converting file "/tmp/tmp.rquJ7pBVhW/00000.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00003.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00007.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00005.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00009.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00004.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00006.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00002.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00001.epoch"
Converting file "/tmp/tmp.rquJ7pBVhW/00008.epoch"

[nc@lorien:~/topos/iohk/ouroboros-network]$ ./ouroboros-consensus/.stack-work/dist/x86_64-linux-nix/Cabal-2.4.0.1/build/byron-db-converter/byron-db-converter v
alidate --configFile $(pwd)/mainnet-genesis.json --genesisHash 5f20df933584822601f9e3f8c024eb5eb252fe8cefb24d1317dc3d432e940ebb --verbose --dbDir /tmp/tmp.hIWE
fz5t3r/
```

### CODE REVIEW

**Comment by mrBliss** [comment]
The first thing I noticed when trying it out:

I was passing in relative directories and it always said:
```
Usage: byron-db-converter convert --epochDir PATH ABS DIR --dbDir PATH ABS DIR
                                  --epochSlots WORD64
..
```
I realised it wanted absolute paths, but the error message isn't very clear. And actually, it should also just accept relative paths :slightly_smiling_face:

**Comment by mrBliss** [comment]
What output should I expect when validating?
I currently get:
```
TraceImmDBEvent ValidatedLastLocation
```
but it is still busy.

Maybe we can print progress after validating each epoch file?

**Comment by nc6** [comment]
Ultimately, we should see it printing the DB tip. But I have so far seen this only with a DB containing only a single epoch. On the other hand, it may be that I haven't waited long enough for larger databases.

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+  -> IO (Either CC.ParseError ())
+convertEpochFile es inFile outDir =
+  let inStream = CC.parseEpochFileWithBoundary es (toFilePath inFile)
+      dbDir = outDir </> [reldir|immutable|]
+      encode = CB.serializeEncoding' . Byron.encodeByronBlock . Byron.ByronBlockOrEBB
+   in do
+        createDirIfMissing True dbDir
+        outFileName <-
+          parseRelFile $ "epoch-" <>
+            drop 2 (toFilePath (filename inFile))
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: Add a comment to `drop 2` that this changes `XXXXX` to `XXX` or just a comment saying that the old format is `XXXXX.epoch` and the new one is `epoch-XXXXX.dat`

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+          initLedgerState = ExtLedgerState
+            { ledgerState = Byron.ByronEBBLedgerState $
+                Byron.ByronLedgerState cvs Seq.empty
+            , ouroborosChainState = Seq.empty
+            }
+          protocolMagicId = CB.unAnnotated . getAProtocolMagicId $ CC.Genesis.configProtocolMagic cfg
+          epochSlots = CC.Genesis.configEpochSlots cfg
+          securityParam = SecurityParam $ CC.unBlockCount k
+          k = CC.Genesis.configK cfg
+          args registry =
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: Using a a few helpers (just in this `where` block is fine) would make it more readable, e.g., a helper for making the `args`, a helper for making the `NodeConfig` (separate one for `PBftNodeConfig` and `ByronConfig`?), a helper for the `Tracer`.

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+              , ChainDB.cdbDecodeBlock = Byron.decodeByronBlock epochSlots
+              , ChainDB.cdbDecodeChainState = Byron.decodeByronChainState
+              , ChainDB.cdbDecodeHash = Byron.decodeByronHeaderHash
+              , ChainDB.cdbDecodeLedger = Byron.decodeByronLedgerState
+              , ChainDB.cdbEncodeBlock = Byron.encodeByronBlock
+              , ChainDB.cdbEncodeChainState = Byron.encodeByronChainState
+              , ChainDB.cdbEncodeHash = Byron.encodeByronHeaderHash
+              , ChainDB.cdbEncodeLedger = Byron.encodeByronLedgerState
+              , -- Policy
+              ChainDB.cdbValidation = ImmDB.ValidateAllEpochs
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: My OCD prefers:
```haskell
, ChainDB.cdbEncodeLedger = Byron.encodeByronLedgerState
-- Policy
, ChainDB.cdbValidation = ImmDB.ValidateAllEpochs
```

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+              , ChainDB.cdbBlocksPerFile = 10
+              , ChainDB.cdbMemPolicy = LgrDB.defaultMemPolicy securityParam
+              , ChainDB.cdbDiskPolicy = LgrDB.defaultDiskPolicy securityParam 20000
+              , -- Integration
+              ChainDB.cdbNodeConfig = WithEBBNodeConfig $ EncNodeConfig
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: Would it be easier to reuse `protocolInfoByron` here? https://github.com/input-output-hk/ouroboros-network/blob/5a9bd9c8393a079d6f95ba66a50640135f8fdcf2/ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs#L114
Also see
https://github.com/input-output-hk/ouroboros-network/blob/e8849236663a9c1e0f9f81e3915f1b612e0a03df/ouroboros-consensus/src/Ouroboros/Consensus/Protocol.hs#L80

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                  , pbftSecrets = error "Attempt to use PBFT secrets in DB validation!"
+                  }
+                }
+              , ChainDB.cdbEpochSize = const (return . EpochSize . unEpochSlots $ epochSlots)
+              , ChainDB.cdbIsEBB = \blk -> case Byron.unByronBlockOrEBB blk of
+                  CC.ABOBBlock _    -> Nothing
+                  CC.ABOBBoundary ebb -> Just (CC.boundaryHashAnnotated ebb)
+              , -- Misc
+              ChainDB.cdbTracer = if verbose
+              then
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: Please indent `then` and `else` (with two spaces) to make it clear that this code is part of the RHS of`ChainDB.cdbTracer`

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  = Convert
+      { epochDir :: w ::: Path Abs Dir <?> "Path to the directory containing old epoch files"
+      , dbDir :: w ::: Path Abs Dir <?> "Path to the new database directory"
+      , epochSlots :: w ::: Word64 <?> "Slots per epoch"
+      }
+  | Validate
+      { dbDir :: w ::: Path Abs Dir <?> "Path to the new database directory"
+      , configFile :: w ::: Path Abs File <?> "Configuration file (e.g. mainnet-genesis.json)"
+      , systemStart :: w ::: Maybe UTCTime <?> "System start time"
+      , requiresNetworkMagic :: w ::: Bool <?> "Expecto patronum?"
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: :smile:

File: ouroboros-consensus/tools/db-convert/Main.hs (hunk 3 above)
Comment (nc6) [reply ouroboros-consensus/tools/db-convert/Main.hs]: As does mine! I've been testing `ormolu`, which I mostly like; this is the one thing I don't!

Comment (mrBliss) [reply ouroboros-consensus/tools/db-convert/Main.hs]: That's the problem I have with formatting tools, there's always at least one thing it does that I don't like :slightly_smiling_face:

**Comment by mrBliss** [comment]
The relative paths don't seem to work:
```haskell
$ stack exec byron-db-converter -- convert --epochDir ../cardano-mainnet-mirror/epochs/ --dbDir converted/ --epochSlots 21600
byron-db-converter: InvalidAbsDir "/home/thomas/work/ouroboros-network/../cardano-mainnet-mirror/epochs/"
```

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+main :: IO ()
+main = do
+  (cmd :: Args Unwrapped) <- unwrapRecord "Byron DB converter"
+  case cmd of
+    Convert {epochDir, dbDir, epochSlots} -> do
+      dbDir' <- parseAbsDir =<< makeAbsolute dbDir
+      (_, files) <- listDir =<< parseAbsDir =<< makeAbsolute epochDir
+      let epochFiles = filter (\f -> fileExtension f == ".epoch") files
+      putStrLn $ "Writing to " ++ show dbDir
+      for_ epochFiles $ \f -> do
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: `sort` the `epochFiles`, that gives a more intuitive output.

File: ouroboros-consensus/tools/db-convert/Main.hs (hunk 3 above)
Comment (mrBliss) [reply ouroboros-consensus/tools/db-convert/Main.hs]: Hmm, can you fix this? And run `stylish-haskell` for uniformity's sake :slightly_smiling_face:

File: ouroboros-consensus/tools/db-convert/Main.hs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        , ChainDB.cdbEncodeBlock = Byron.encodeByronBlock
+        , ChainDB.cdbEncodeChainState = Byron.encodeByronChainState
+        , ChainDB.cdbEncodeHash = Byron.encodeByronHeaderHash
+        , ChainDB.cdbEncodeLedger = Byron.encodeByronLedgerState
+        , -- Policy
+        ChainDB.cdbValidation = ImmDB.ValidateAllEpochs
+        , ChainDB.cdbBlocksPerFile = 10
+        , ChainDB.cdbMemPolicy = LgrDB.defaultMemPolicy securityParam
+        , ChainDB.cdbDiskPolicy = LgrDB.defaultDiskPolicy securityParam 20000
+        , -- Integration
```
Comment (mrBliss) [inline ouroboros-consensus/tools/db-convert/Main.hs]: Can you fix this line too? And maybe run `stylish-haskell` for uniformity's sake :slightly_smiling_face:

Comment (nc6) [reply ouroboros-consensus/tools/db-convert/Main.hs]: `stylish-haskell` still relies on a version which isn't in nixpkgs, and I'm too stubborn to install...

**Pull request merged**